### PR TITLE
http: Lazily build http/tcp stacks

### PR DESCRIPTION
**Pull request by olix0r**
Currently, the `DetectHttp` layer always builds both HTTP and TCP
stacks, even though only one is needed. Furthermore, we lose the
information about which protocol was detected--it's never passed into
the http stack (and instead must be inferred from individual requests).

This change modifies the `DetectHttp` layer to only build inner services
as needed, after the protocol has been detected.

Furthermore, outbound caching has been moved up from the tcp balancer to
wrap the HTTP detection layer. This was necessary so that the tcp
balancer could be built instantaneously as a NewService while still
being cloneable. This means that, for each outbound target ip:port, we
will reuse HTTP and TCP stacks.

However, HTTP routing remains dependent on per-request metadata. This is
expected to change in a followup.

### CODE REVIEW

File: linkerd/app/outbound/src/lib.rs [hunk 1]
```diff
@@ -485,37 +487,32 @@ impl Config {
...
-                        .push_spawn_buffer_with_idle_timeout(buffer_capacity, cache_max_idle_age),
-                ),
+            .push_on_response(
+                svc::layers()
+                    .push_failfast(dispatch_timeout)
+                    .push_spawn_buffer_with_idle_timeout(buffer_capacity, cache_max_idle_age),
             )
-            .into_make_service()
-            .spawn_buffer(buffer_capacity)
-            .instrument(|_: &_| info_span!("tcp"))
```
Comment (hawkw) [inline linkerd/app/outbound/src/lib.rs:517]: nit: do we still want the instrument layer here?

**Review by hawkw — APPROVED** [review]
this change looks great to me --- i commented on a few minor nits.

File: linkerd/app/outbound/src/lib.rs (hunk 1 above)
Comment (olix0r) [reply linkerd/app/outbound/src/lib.rs:517]: the tcp scope is now set in detect-http. it's probably better to stop doing that and keep them in here

File: linkerd/proxy/http/src/detect.rs [hunk 2]
```diff
@@ -155,50 +132,85 @@ where
...
+                debug!("Handling as H2");
+                let h2 = if let Some(svc) = self.h2.clone() {
+                    svc
+                } else {
+                    let svc = self
+                        .new_http
+                        .new_service((HttpVersion::H2, self.target.clone()));
+                    self.h2 = Some(svc.clone());
+                    svc
+                };
```
[the remark this answers is not in the document]
Comment (olix0r) [reply linkerd/proxy/http/src/detect.rs:179]: That doesn't work due to borrows on self.

File: linkerd/proxy/http/src/detect.rs [hunk 3]
```diff
@@ -155,50 +132,85 @@ where
...
+                debug!("Handling as HTTP");
+                let http1 = if let Some(svc) = self.http1.clone() {
+                    svc
+                } else {
+                    let svc = self
+                        .new_http
+                        .new_service((HttpVersion::Http1, self.target.clone()));
+                    self.http1 = Some(svc.clone());
+                    svc
+                };
```
[the remark this answers is not in the document]
Comment (olix0r) [reply linkerd/proxy/http/src/detect.rs:147]: That doesn't work due to borrows on self.

File: linkerd/proxy/http/src/detect.rs [hunk 4]
```diff
@@ -37,9 +37,13 @@ pub struct DetectHttp<F, H> {
...
-pub struct AcceptHttp<F, H> {
-    tcp: F,
-    http: H,
+pub struct AcceptHttp<T, F: NewService<T>, H: NewService<(HttpVersion, T)>> {
+    target: T,
+    new_tcp: F,
+    tcp: Option<F::Service>,
+    new_http: H,
+    http1: Option<H::Service>,
+    h2: Option<H::Service>,
```
[the remark this answers is not in the document]
Comment (olix0r) [reply linkerd/proxy/http/src/detect.rs:46]: It corresponds to the `Version` variant

**Pull request merged**